filters/DBSCANFilter.cpp

```diff
@@ -0,0 +1,148 @@
+// https://github.com/intel-isl/Open3D/pull/1038.
+
+#include "DBSCANFilter.hpp"
+
+#include <pdal/KDIndex.hpp>
+
+#include <string>
+#include <unordered_set>
+
+namespace pdal
+{
+
+static StaticPluginInfo const s_info{
+    "filters.dbscan", "DBSCAN Clustering.",
+    "http://pdal.io/stages/filters.dbscan.html"};
+
+CREATE_STATIC_STAGE(DBSCANFilter, s_info)
+
+std::string DBSCANFilter::getName() const
+{
+    return s_info.name;
+}
+
+void DBSCANFilter::addArgs(ProgramArgs& args)
+{
+    args.add("min_points", "Min points per cluster", m_minPoints,
+             static_cast<uint64_t>(6));
+    args.add("eps", "Epsilon", m_eps, 1.0);
+}
+
+void DBSCANFilter::addDimensions(PointLayoutPtr layout)
+{
+    m_cluster =
+        layout->registerOrAssignDim("ClusterID", Dimension::Type::Signed64);
+}
+
+void DBSCANFilter::filter(PointView& view)
+{
+    // Construct 3D KDIndex for radius search.
+    KD3Index& kdi = view.build3dIndex();
+
+    // First pass through point cloud precomputes neighbor indices and
+    // initializes ClusterID to -2.
+    std::vector<std::vector<PointId>> neighbors(view.size());
+    for (PointId idx = 0; idx < view.size(); ++idx)
+    {
+        neighbors[idx] = kdi.radius(idx, m_eps);
+        view.setField(m_cluster, idx, -2);
+    }
+
+    // Second pass through point cloud performs DBSCAN clustering.
+    int64_t cluster_label = 0;
+    for (PointId idx = 0; idx < view.size(); ++idx)
+    {
+        // Point has already been labeled, so move on to next.
+        if (view.getFieldAs<int64_t>(m_cluster, idx) != -2)
+            continue;
+
+        // Density of the neighborhood does not meet minimum number of points
+        // constraint, label as noise.
+        if (neighbors[idx].size() < m_minPoints)
+        {
+            view.setField(m_cluster, idx, -1);
+            continue;
+        }
+
+        // Initialize some bookkeeping to track which neighbors have been
+        // considered for addition to the current cluster.
+        std::unordered_set<PointId> neighbors_next(neighbors[idx].begin(),
+                                                   neighbors[idx].end());
+        std::unordered_set<PointId> neighbors_visited;
+        neighbors_visited.insert(idx);
+
+        // Unlabeled point encountered; assign cluster label.
+        view.setField(m_cluster, idx, cluster_label);
+
+        // Consider all neighbors.
+        while (!neighbors_next.empty())
+        {
+            // Select first neighbor and move it to the visited set.
+            PointId p = *neighbors_next.begin();
+            neighbors_next.erase(neighbors_next.begin());
+            neighbors_visited.insert(p);
+
+            // Reassign cluster label to neighbor previously marked as noise.
+            if (view.getFieldAs<int64_t>(m_cluster, p) == -1)
+                view.setField(m_cluster, p, cluster_label);
+
+            // Neighbor has already been labeled, so move on to next.
```

Comment:
Do you mean this as an else case?  If not, perhaps better just to fetch the value once from the point view and set when we're done.

Reply:
Maybe this can be rewritten as

```c++
int64_t cluster_val = view.getFieldAs<int64_t>(m_cluster, p) == -1);
if (cluster_val ==-1)
{
    view.setField(m_cluster, p, cluster_label);
    continue;
}
else if (cluster_val != -2)
{
    continue;
}
```

Not sure if that's any clearer, and I'd need to double check the logic. Point being that as written it correctly sets a cluster label if it was noise before, and regardless of whether or not it was originally labelled, or was relabeled from noise, as long as it now has a label, we hit the continue at line 126.